### PR TITLE
feat: ensure no low cardinality snapshot source in any deployment

### PR DESCRIPTION
**Pull request by pauldambra**
I initially created the snapshot source column as low cardinality nullable string, since we know it can be null or one of a small number of values... ClickHouse really didn't like that. 

We've cleared this up on CH cloud, but there could be hobby deployments, dev deployments that have low cardinality on some of them and, since we're not using the values in the column, it should be safe to drop and recreate the column so we know it has no LowCardinality references anywhere

### CODE REVIEW

**Comment by benjackwhite** [comment]
> maybe @benjackwhite has time to test this too
> 
> ```
> SELECT
>     type ilike '%LowCardinality%' AS has_low_cardinality,
>     type AS column_type,
>     name AS column_name,
>     table
> FROM system.columns
> WHERE
>     database = 'default' AND
>     lower(name) = 'snapshot_source'
> ORDER BY table, name;
> ```
> 
> running that _should_ print some columns with LowCardinality still in the definition.
> 
> Then run the migration in this branch
> 
> Everything should still work _and_ re-running the above query should show no columns with low cardinality in the definition

This all worked as you say except that the migrate command timed out the first time (on the 5th command) and the second time on the 1st command...

And now my migrations are completely stuck...

**Comment by pauldambra** [comment]
@daibhin we nuked Ben's environment and the migration then worked... that's not great if this breaks a self-hosted deploy and the only advice is to delete everything 🙈 

can you test locally and see what happens for you?

**Comment by daibhin** [comment]
Running the first time
```
┌─has_low_cardinality─┬─column_type───────────────────────────────────────────────────────────────────────┬─column_name─────┬─table──────────────────────────┐
│                   1 │ LowCardinality(Nullable(String))                                                  │ snapshot_source │ kafka_session_replay_events    │
│                   1 │ AggregateFunction(argMin, LowCardinality(Nullable(String)), DateTime64(6, 'UTC')) │ snapshot_source │ session_replay_events          │
│                   1 │ AggregateFunction(argMin, LowCardinality(Nullable(String)), DateTime64(6, 'UTC')) │ snapshot_source │ session_replay_events_mv       │
│                   1 │ AggregateFunction(argMin, LowCardinality(Nullable(String)), DateTime64(6, 'UTC')) │ snapshot_source │ sharded_session_replay_events  │
│                   1 │ AggregateFunction(argMin, LowCardinality(Nullable(String)), DateTime64(6, 'UTC')) │ snapshot_source │ writable_session_replay_events │
└─────────────────────┴───────────────────────────────────────────────────────────────────────────────────┴─────────────────┴────────────────────────────────┘

5 rows in set. Elapsed: 0.005 sec.
```

Ran the migration - all went well
```
2024-02-01T12:50:21.728088Z [info     ] Applying migration 0052_session_replay_source_ensure_high_cardinality... [migrations] pid=11665 tid=8016367616
2024-02-01T12:50:21.728171Z [info     ]     Executing python operation run_sql [migrations] pid=11665 tid=8016367616
2024-02-01T12:50:21.850276Z [info     ]     Executing python operation run_sql [migrations] pid=11665 tid=8016367616
2024-02-01T12:50:25.141114Z [info     ]     Executing python operation run_sql [migrations] pid=11665 tid=8016367616
2024-02-01T12:50:25.257037Z [info     ]     Executing python operation run_sql [migrations] pid=11665 tid=8016367616
2024-02-01T12:50:25.380275Z [info     ]     Executing python operation run_sql [migrations] pid=11665 tid=8016367616
2024-02-01T12:50:25.500814Z [info     ]     Executing python operation run_sql [migrations] pid=11665 tid=8016367616
2024-02-01T12:50:25.609655Z [info     ]     Executing python operation run_sql [migrations] pid=11665 tid=8016367616
2024-02-01T12:50:25.722659Z [info     ]     Executing python operation run_sql [migrations] pid=11665 tid=8016367616
2024-02-01T12:50:25.841519Z [info     ]     Executing python operation run_sql [migrations] pid=11665 tid=8016367616
2024-02-01T12:50:25.965506Z [info     ]     Executing python operation run_sql [migrations] pid=11665 tid=8016367616
✅ Migration successful
```

Looks like the columns are gone
```
┌─has_low_cardinality─┬─column_type───────────────────────────────────────────────────────┬─column_name─────┬─table──────────────────────────┐
│                   0 │ Nullable(String)                                                  │ snapshot_source │ kafka_session_replay_events    │
│                   0 │ AggregateFunction(argMin, Nullable(String), DateTime64(6, 'UTC')) │ snapshot_source │ session_replay_events          │
│                   0 │ AggregateFunction(argMin, Nullable(String), DateTime64(6, 'UTC')) │ snapshot_source │ session_replay_events_mv       │
│                   0 │ AggregateFunction(argMin, Nullable(String), DateTime64(6, 'UTC')) │ snapshot_source │ sharded_session_replay_events  │
│                   0 │ AggregateFunction(argMin, Nullable(String), DateTime64(6, 'UTC')) │ snapshot_source │ writable_session_replay_events │
└─────────────────────┴───────────────────────────────────────────────────────────────────┴─────────────────┴────────────────────────────────┘

5 rows in set. Elapsed: 0.004 sec.
```

**Comment by daibhin** [comment]
Random question @pauldambra: I queried CH by running `clickhouse client` on Exec tab of the container in Docker UI. Is there a better way to connect to a CH shell locally?

**Comment by pauldambra** [comment]
>  I queried CH by running clickhouse client on Exec tab of the container in Docker UI. Is there a better way to connect to a CH shell locally?

I've seen people using "table plus"
I use pycharm which has a DB explorer built in
JetBrains also have a DB thingy called DataGrip iirc

I'm not sure what else people use

File: posthog/session_recordings/sql/session_replay_event_migrations_sql.py [hunk 1]
```diff
@@ -114,3 +114,49 @@
     table_name=SESSION_REPLAY_EVENTS_DATA_TABLE(),
     cluster=settings.CLICKHOUSE_CLUSTER,
 )
+
+ALTER_SESSION_REPLAY_DROP_SOURCE_COLUMN = """
+    ALTER TABLE {table_name} on CLUSTER '{cluster}'
+    DROP COLUMN IF EXISTS snapshot_source
+"""
+
+DROP_SOURCE_DISTRIBUTED_SESSION_REPLAY_EVENTS_TABLE_SQL = lambda: ALTER_SESSION_REPLAY_DROP_SOURCE_COLUMN.format(
```
Comment (fuziontech) [inline posthog/session_recordings/sql/session_replay_event_migrations_sql.py:123]: I really don't like the name of this `DROP_SOURCE_DISTRIBUTED_SESSION_REPLAY_EVENTS_TABLE_SQL` It reads like a drop table and doesn't reference that it's just a column. When I first looked at the migration above I was worried that we were dropping the whole table

File: posthog/clickhouse/migrations/0052_session_replay_source_ensure_high_cardinality.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    # now we can drop the snapshot source column, it's currently unused so safe to drop,
+    # and in some deployments might still have low cardinality as part of the type
+    # this should be fine but CH really doesn't like it
+    run_sql_with_exceptions(DROP_SOURCE_DISTRIBUTED_SESSION_REPLAY_EVENTS_TABLE_SQL()),
+    run_sql_with_exceptions(DROP_SOURCE_WRITABLE_SESSION_REPLAY_EVENTS_TABLE_SQL()),
+    run_sql_with_exceptions(DROP_SOURCE_SESSION_REPLAY_EVENTS_TABLE_SQL()),
+    # now we can add the snapshot source columns back without the low cardinality
+    run_sql_with_exceptions(ADD_HIGH_CARDINALITY_SOURCE_DISTRIBUTED_SESSION_REPLAY_EVENTS_TABLE_SQL()),
+    run_sql_with_exceptions(ADD_HIGH_CARDINALITY_SOURCE_WRITABLE_SESSION_REPLAY_EVENTS_TABLE_SQL()),
+    run_sql_with_exceptions(ADD_HIGH_CARDINALITY_SOURCE_SESSION_REPLAY_EVENTS_TABLE_SQL()),
```
Comment (fuziontech) [inline posthog/clickhouse/migrations/0052_session_replay_source_ensure_high_cardinality.py:31]: I'd just make this more clear that these are `column` operations, the rest looks ✅

File: posthog/session_recordings/sql/session_replay_event_migrations_sql.py (hunk 1 above)
Comment (pauldambra) [reply posthog/session_recordings/sql/session_replay_event_migrations_sql.py:123]: no, you were copying and pasting previous blocks and using find and replace 🙈

**Pull request closed without merging**